tool/tctl/common/device/resource_test.go

```diff
@@ -0,0 +1,110 @@
+// Copyright 2023 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package device
+
+import (
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	kyaml "k8s.io/apimachinery/pkg/util/yaml"
+
+	devicepb "github.com/gravitational/teleport/api/gen/proto/go/teleport/devicetrust/v1"
+	"github.com/gravitational/teleport/lib/defaults"
+	"github.com/gravitational/teleport/lib/services"
+)
+
+// TestUnmarshalDevice tests that devices can be successfully
+// unmarshalled from YAML and JSON.
+func TestUnmarshalDevice(t *testing.T) {
+	t.Parallel()
+	for _, tc := range []struct {
+		desc          string
+		input         string
+		errorContains string
+		expected      *devicepb.Device
+	}{
+		{
+			desc: "success",
+			input: `---
+kind: device
+version: v1
+metadata:
+  name: xaa
+spec:
+  api_version: v1
+  id: xaa
+  asset_tag: mymachine
+  os_type: 2
```

Comment:
Could we build the resource so OS type matches the inputs to `tctl devices add` (and similar commands)? Ie, the OS name instead of the numeric enum value.

Reply:
I've set it to the be the string representation of the protos now. Do you think we should keep yet another set of maps containing lower case "resource friendly" names or keep as is?

Reply:
The same goes for the enrollment status enum btw.

Reply:
I think a pair of mapping functions under lib/devicetrust, in the inverse direction of the "Friendly$Enum" ones, is best here.

The problem with using the proto enum maps is that it "leaks" proto conventions like prefixing type names, so users would have to write "OS_TYPE_LINUX", which is a lot weirder than writing "LINUX" (and so on). Whether case-sensitive or not I'll leave up to you, although following `tctl devices` is likely a good call. Not sure if we have conventions for this kind of thing either.

Reply:
Note: enrollment status, unlike the OS, is never filled-in by the user, so there's less of a worry there. It may makes things a bit easier (you may be able to get away without the string->enum logic, enum->string we already have).